tests/integration/v3_watch_restore_test.go

```diff
@@ -154,3 +186,89 @@ func TestV3WatchRestoreSnapshotUnsync(t *testing.T) {
 		}
 	}
 }
+
+type lineCountExpecter struct {
+	doneOnce sync.Once
+	doneCh   chan struct{}
+
+	content string
+	count   int64
+	seen    int64
+}
+
+func newlineCountExpecter(interestingContent string, expectedCount int64) *lineCountExpecter {
+	return &lineCountExpecter{
+		doneCh:  make(chan struct{}),
+		content: interestingContent,
+		count:   expectedCount,
+	}
+}
+
+func (le *lineCountExpecter) Notify(log string) {
+	if !strings.Contains(log, le.content) {
+		return
+	}
+
+	if atomic.AddInt64(&le.seen, 1) >= le.count {
+		le.doneOnce.Do(func() {
+			close(le.doneCh)
+		})
+	}
+}
+
+func (le *lineCountExpecter) wait(timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(context.TODO(), timeout)
+	defer cancel()
+
+	select {
+	case <-le.doneCh:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+type testingLogfSubscriber interface {
+	Notify(log string)
+}
+
+// testingLogfMonitor is to monitor t.Logf output.
+type testingLogfMonitor struct {
```

Comment:
https://github.com/etcd-io/etcd/blob/f7af6b64bafcc32071b9867ffc7447e3c66111ae/tests/framework/integration/cluster.go#L746

I think we can extend the zaptest.Core with tee function. It can be aligned with AssertProcessLogs for process expecter~